firedrake/mg/ufl_utils.py

```diff
@@ -126,35 +126,17 @@ def coarsen_bc(bc, self, coefficient_mapping=None):
     return type(bc)(V, val, subdomain)
 
 
-@coarsen.register(firedrake.functionspaceimpl.FiredrakeDualSpace)
-@coarsen.register(firedrake.functionspaceimpl.FunctionSpace)
-@coarsen.register(firedrake.functionspaceimpl.WithGeometry)
+@coarsen.register(firedrake.functionspaceimpl.WithGeometryBase)
 def coarsen_function_space(V, self, coefficient_mapping=None):
     if hasattr(V, "_coarse"):
         return V._coarse
-    fine = V
-    indices = []
-    while True:
-        if V.index is not None:
-            indices.append(V.index)
-        if V.component is not None:
-            indices.append(V.component)
-        if V.parent is not None:
-            V = V.parent
-        else:
-            break
-
-    mesh = self(V.mesh(), self)
-
-    if isinstance(V, firedrake.functionspaceimpl.FiredrakeDualSpace):
-        V = firedrake.functionspace.DualSpace(mesh, V.ufl_element())
-    else:
-        V = firedrake.FunctionSpace(mesh, V.ufl_element())
 
-    for i in reversed(indices):
-        V = V.sub(i)
-    V._fine = fine
-    fine._coarse = V
+    V_fine = V
+    mesh_coarse = self(V_fine.mesh(), self)
+    name = f"coarse_{V.name}" if V.name else None
+    V_coarse = V_fine.reconstruct(mesh=mesh_coarse, name=name)
+    V_coarse._fine = V_fine
+    V_fine._coarse = V_coarse
 
     # FIXME: This replicates some code from dmhooks.coarsen, but we
```

Comment:
@wence- I believe we addressed this over here: https://github.com/firedrakeproject/firedrake/blob/2ae2ce8c6289f86b8e1aac35ad254313b1d6761c/firedrake/variational_solver.py#L280

Can we remove this block of code?